src/relative_duration/parse.rs

```diff
@@ -0,0 +1,191 @@
+use super::RelativeDuration;
+use chrono::Duration;
+
+fn dhms_to_duration(days: i32, hours: i32, minutes: i32, seconds: i32) -> Duration {
+    Duration::seconds((((days * 24 + hours) * 60 + minutes) * 60 + seconds) as i64)
+}
+
+fn get_terminated(input: &str, terminator: char) -> Result<(&str, i32), String> {
+    if let Some((int_string, remainder)) = input.split_once(terminator) {
+        let int = int_string
+            .parse::<i32>()
+            .map_err(|_| format!("{} is not a valid i32", int_string))?;
+        Ok((remainder, int))
+    } else {
+        Ok((input, 0))
+    }
+}
+
+fn parse_datespec(datespec: &str) -> Result<(i32, i32, i32), String> {
+    let (remainder, years) = get_terminated(datespec, 'Y')?;
+    let (remainder, months) = get_terminated(remainder, 'M')?;
+    let (remainder, weeks) = get_terminated(remainder, 'W')?;
+    let (remainder, days) = get_terminated(remainder, 'D')?;
+
+    if !remainder.is_empty() {
+        Err(format!(
+            "trailing characters: {} in datespec: {}",
+            remainder, datespec
+        ))
+    } else {
+        Ok((years, months, (weeks * 7) + days))
+    }
+}
+
+fn parse_timespec(timespec: &str) -> Result<(i32, i32, i32), String> {
+    let (remainder, hours) = get_terminated(timespec, 'H')?;
+    let (remainder, mins) = get_terminated(remainder, 'M')?;
+    let (remainder, secs) = get_terminated(remainder, 'S')?;
+
+    if !remainder.is_empty() {
+        Err(format!(
+            "trailing characters: {} in timespec: {}",
+            remainder, timespec
+        ))
+    } else {
+        Ok((hours, mins, secs))
+    }
+}
+
+fn format_spec(nums: [i64; 3], chars: [char; 3]) -> String {
+    nums.iter()
+        .zip(chars)
+        .filter(|x| *x.0 != 0)
+        .map(|x| format!("{}{}", x.0, x.1))
+        .reduce(|acc, x| acc + &x)
+        .unwrap_or_else(|| "".to_string())
+}
```

Comment:
I don't think you need reduce or unwrap_or_else here, you can just write
```rust
    .collect()
```
to collect an iterator of &str into String without making the intermediate allocations.

Reply:
This has ended up outdated, since I had to change the formatting implementation quite heavily to support quirks introduced by nanosecond support.